README.md

```diff
@@ -1,2 +1,108 @@
+  - **Room placement algorithm:** Randomly placing rooms into dungeon using Kruskal algorithm
+- **Error handling:** Validate configurations and provides descriptive errors for invalid parameters
+
+## How It Works
+
+### DungeonBuilder API
+Central part of the project is a **DungeonBuilder** trait, which allows to define dungeon building procedure.
+It provides DungeonBuildConfig and implementor of the trait should return **Dungeon** which represents a 2d map of tiles.
+
+```Rust
+pub trait DungeonBuilder {
+    fn build(self, build_config: DungeonBuildConfig) -> Result<Dungeon, DungeonBuildError>;
+}
+```
+
+```Rust
+#[repr(u32)]
+pub enum TileType {
+    Door = 1,
+    Wall = 4,
+    Floor = 5,
+}
+
+pub struct Dungeon {
+    pub map: Vec<Vec<TileType>>,
+}
+```
+**TileType** is represented by u32 so it is easier to map to tile in the rendering engine like bevy.
+
+### DungeonConfigBuilder<BuilderAlgorithm>
+
+Another important part of the library is DungeonConfigBuilder - it allows to configure all necessary parameters that are required to build dungeon and it allows to configure BuilderAlgorithm which is defined as parameter with trait constraints.
+
+### Algorithms
+
+#### Partition (BSP) Algorithm
+- Process:
+  - Recursively splits the dungeon area into smaller rectangles (leaf nodes) using Binary Space Partitioning (BSP).
+  - Places rooms within these leaf nodes.
+  - Connects rooms using corridors between the centers of rooms in sibling leaf nodes.
+- Advantages:
+  - Creates a more structured dungeon layout.
+  - Rooms are well-distributed across the dungeon.
+
+#### Room Placement Algorithm
+- Process:
+  - Randomly places rooms on the map without overlapping.
+  - Uses a connectivity graph to connect rooms using corridors.
+  - Employs Kruskal's algorithm to generate a Minimum Spanning Tree (MST) for connecting rooms efficiently.
+- Advantages:
+  - Creates a more organic and random dungeon layout.
+  - Can result in complex and non-linear dungeon designs.
+
+## How to use it
+
+```Rust
+let dungeon = DungeonConfigBuilder::new()
+        .dungeon_size(DungeonSize { width: w, height: h })
+        .room_size(RoomSize { min_room_size: min_size, max_room_size: max_size })
+        .build_algorithm(BinaryPartitionBuilder)
+        .should_place_doors(false)
+        .build()
+        .expect("Failed to build dungeon");
+let map = dungeon.map;
+```
+
+Map is a 2d array which can be used in your favority game engine to render your tiles of choice.
+**DungeonConfigBuilder** provides default values, but you can configure - dungeon size, room min and max size. Select build algorithm and set if you wnat to place doors or not.
+While building your dungeon - different error can happen, so you should handle errors, they are represented by Enum
+
+```Rust
+#[Error(desc = "Dungeon generation error", fmt = debug)]
+pub enum DungeonBuildError {
+    #[error(desc = "Size of map is not valid", fmt = display)]
```

Comment:
Small proposal: Enhance error message by including width and height for more informative feedback.

Something like this:

```Rust
#[error(desc = "Size of map is not valid: width = {0}, height = {1}", fmt = display)]
InvalidSize(usize, usize),
```